include/capstone/riscv.h

```diff
@@ -0,0 +1,428 @@
+	RISCV_REG_F12_64, 
+	RISCV_REG_F13_64, 
+	RISCV_REG_F14_64, 
+	RISCV_REG_F15_64, 
+	RISCV_REG_F16_64, 
+	RISCV_REG_F17_64, 
+	RISCV_REG_F18_64, 
+	RISCV_REG_F19_64,  
+	RISCV_REG_F20_64,
+	RISCV_REG_F21_64,
+	RISCV_REG_F22_64,
+	RISCV_REG_F23_64,
+	RISCV_REG_F24_64,
+	RISCV_REG_F25_64,
+	RISCV_REG_F26_64,
+	RISCV_REG_F27_64,
+	RISCV_REG_F28_64,
+	RISCV_REG_F29_64,
+	RISCV_REG_F30_64,
+	RISCV_REG_F31_64,
+
+	RISCV_REG_ENDING,	// <-- mark the end of the list or registers
+} riscv_reg;
+
+//> RISCV instruction
+typedef enum riscv_insn {
+	RISCV_INS_INVALID = 0,
+
+	RISCV_INS_ADD,
+	RISCV_INS_ADDI,
+	RISCV_INS_ADDIW,
+	RISCV_INS_ADDW,
+	RISCV_INS_AMOADD_D,
+	RISCV_INS_AMOADD_D_AQ,
+	RISCV_INS_AMOADD_D_AQ_RL,
+	RISCV_INS_AMOADD_D_RL,
+	RISCV_INS_AMOADD_W,
+	RISCV_INS_AMOADD_W_AQ,
+	RISCV_INS_AMOADD_W_AQ_RL,
+	RISCV_INS_AMOADD_W_RL,
+	RISCV_INS_AMOAND_D,
+	RISCV_INS_AMOAND_D_AQ,
+	RISCV_INS_AMOAND_D_AQ_RL,
+	RISCV_INS_AMOAND_D_RL,
+	RISCV_INS_AMOAND_W,
+	RISCV_INS_AMOAND_W_AQ,
+	RISCV_INS_AMOAND_W_AQ_RL,
+	RISCV_INS_AMOAND_W_RL,
+	RISCV_INS_AMOMAXU_D,
+	RISCV_INS_AMOMAXU_D_AQ,
+	RISCV_INS_AMOMAXU_D_AQ_RL,
+	RISCV_INS_AMOMAXU_D_RL,
+	RISCV_INS_AMOMAXU_W,
+	RISCV_INS_AMOMAXU_W_AQ,
+	RISCV_INS_AMOMAXU_W_AQ_RL,
+	RISCV_INS_AMOMAXU_W_RL,
+	RISCV_INS_AMOMAX_D,
+	RISCV_INS_AMOMAX_D_AQ,
+	RISCV_INS_AMOMAX_D_AQ_RL,
+	RISCV_INS_AMOMAX_D_RL,
+	RISCV_INS_AMOMAX_W,
+	RISCV_INS_AMOMAX_W_AQ,
+	RISCV_INS_AMOMAX_W_AQ_RL,
+	RISCV_INS_AMOMAX_W_RL,
+	RISCV_INS_AMOMINU_D,
+	RISCV_INS_AMOMINU_D_AQ,
+	RISCV_INS_AMOMINU_D_AQ_RL,
+	RISCV_INS_AMOMINU_D_RL,
+	RISCV_INS_AMOMINU_W,
+	RISCV_INS_AMOMINU_W_AQ,
+	RISCV_INS_AMOMINU_W_AQ_RL,
+	RISCV_INS_AMOMINU_W_RL,
+	RISCV_INS_AMOMIN_D,
+	RISCV_INS_AMOMIN_D_AQ,
+	RISCV_INS_AMOMIN_D_AQ_RL,
+	RISCV_INS_AMOMIN_D_RL,
+	RISCV_INS_AMOMIN_W,
+	RISCV_INS_AMOMIN_W_AQ,
+	RISCV_INS_AMOMIN_W_AQ_RL,
+	RISCV_INS_AMOMIN_W_RL,
+	RISCV_INS_AMOOR_D,
+	RISCV_INS_AMOOR_D_AQ,
+	RISCV_INS_AMOOR_D_AQ_RL,
+	RISCV_INS_AMOOR_D_RL,
+	RISCV_INS_AMOOR_W,
+	RISCV_INS_AMOOR_W_AQ,
+	RISCV_INS_AMOOR_W_AQ_RL,
+	RISCV_INS_AMOOR_W_RL,
+	RISCV_INS_AMOSWAP_D,
+	RISCV_INS_AMOSWAP_D_AQ,
+	RISCV_INS_AMOSWAP_D_AQ_RL,
+	RISCV_INS_AMOSWAP_D_RL,
+	RISCV_INS_AMOSWAP_W,
+	RISCV_INS_AMOSWAP_W_AQ,
+	RISCV_INS_AMOSWAP_W_AQ_RL,
+	RISCV_INS_AMOSWAP_W_RL,
+	RISCV_INS_AMOXOR_D,
+	RISCV_INS_AMOXOR_D_AQ,
+	RISCV_INS_AMOXOR_D_AQ_RL,
+	RISCV_INS_AMOXOR_D_RL,
+	RISCV_INS_AMOXOR_W,
+	RISCV_INS_AMOXOR_W_AQ,
+	RISCV_INS_AMOXOR_W_AQ_RL,
+	RISCV_INS_AMOXOR_W_RL,
+	RISCV_INS_AND,
+	RISCV_INS_ANDI,
+	RISCV_INS_AUIPC,
+	RISCV_INS_BEQ,
+	RISCV_INS_BGE,
+	RISCV_INS_BGEU,
+	RISCV_INS_BLT,
+	RISCV_INS_BLTU,
+	RISCV_INS_BNE,
+	RISCV_INS_CSRRC,
+	RISCV_INS_CSRRCI,
+	RISCV_INS_CSRRS,
+	RISCV_INS_CSRRSI,
+	RISCV_INS_CSRRW,
+	RISCV_INS_CSRRWI,
+	RISCV_INS_DIV,
+	RISCV_INS_DIVU,
+	RISCV_INS_DIVUW,
+	RISCV_INS_DIVW,
+	RISCV_INS_EBREAK,
+	RISCV_INS_ECALL,
+	RISCV_INS_FADD_D,
+	RISCV_INS_FADD_S,
+	RISCV_INS_FCLASS_D,
+	RISCV_INS_FCLASS_S,
+	RISCV_INS_FCVT_D_L,
+	RISCV_INS_FCVT_D_LU,
+	RISCV_INS_FCVT_D_S,
```

Comment:
my impression is that we can map related instructions into one, like RISCV_INS_FCVT_L_D & RISCV_INS_FCVT_L_S into just RISCV_INS_FCVT_L, or even all related ones into RISCV_INS_FCVT.

correct me if i am wrong (i dont know much about RISCV), but this is how we did with other archs, like we map ADDxxx into just ADD on X86.

Reply:
Is this related to opcode? I have no idea if it could be mapped like x86.

Reply:
If the mapping that you suggest is related to the opcode, then yes, probably instructions with the same opcode can be mapped together. But I am not sure how the mapping is done in x86.

Reply:
@porto703, the mapping is all in xxxMappingInsn.inc file. for X86, you can grep X86_ADD from X86MappingInsn.inc file, and get these lines:

```
X86MappingInsn.inc:     X86_ADD16i16, X86_INS_ADD,
X86MappingInsn.inc:     X86_ADD16mi, X86_INS_ADD,
X86MappingInsn.inc:     X86_ADD16mi8, X86_INS_ADD,
X86MappingInsn.inc:     X86_ADD16mr, X86_INS_ADD,
X86MappingInsn.inc:     X86_ADD16ri, X86_INS_ADD,
X86MappingInsn.inc:     X86_ADD16ri8, X86_INS_ADD,
X86MappingInsn.inc:     X86_ADD16rm, X86_INS_ADD,
X86MappingInsn.inc:     X86_ADD16rr, X86_INS_ADD,
...
```

this is how we map all X86_ADDxxx to X86_INS_ADD, which is the opcode of all ADD instructions, regardless of operand types.

what do you think, should we do the same thing for RISCV?

Reply:
So it seems they are mapped based on the opcode. In RISCV several instructions can be mapped into the same opcode group, but still there are other fields that are used to select the type of operation within the same opcode group. I didn't have the chance to look further into capstone to understand how the mapping was being used, and if this may fit for RISCV. So that is why I didn't include a more refined mapping into the first version that I worked on.
But at first glance, it looks to me that a similar mapping may be done based on the opcode.
Still I would like to see @citypw opinion on this.

Reply:
well, I checked the both x86 and RV manuals a bit. It seems very similar which only mapped to the specific subgroup of opcode. I only tested the "add*" ins:

https://github.com/citypw/capstone/commit/ac4cc0f5b67a42230024121ac3b0b5be847b7751
https://github.com/citypw/capstone/commit/d8d8adb6c2ebd4eb34d9f244c357786135a02d8f

The results of regression test case are the same. I don't know. Maybe we should work toward to cut the mapped ins into fewer ones?

Reply:
> @porto703, the mapping is all in xxxMappingInsn.inc file. for X86, you can grep X86_ADD from X86MappingInsn.inc file, and get these lines:
> 
> ```
> X86MappingInsn.inc:     X86_ADD16i16, X86_INS_ADD,
> X86MappingInsn.inc:     X86_ADD16mi, X86_INS_ADD,
> X86MappingInsn.inc:     X86_ADD16mi8, X86_INS_ADD,
> X86MappingInsn.inc:     X86_ADD16mr, X86_INS_ADD,
> X86MappingInsn.inc:     X86_ADD16ri, X86_INS_ADD,
> X86MappingInsn.inc:     X86_ADD16ri8, X86_INS_ADD,
> X86MappingInsn.inc:     X86_ADD16rm, X86_INS_ADD,
> X86MappingInsn.inc:     X86_ADD16rr, X86_INS_ADD,
> ...
> ```
> this is how we map all X86_ADDxxx to X86_INS_ADD, which is the opcode of all ADD instructions, regardless of operand types.
> 
> what do you think, should we do the same thing for RISCV?

In riscv this is not possible, it is the only reason tell the CPU which is 32bit or 64bit.
for example, lw/ld can't merge to load. it only tells the difference according to the w/d.
not like X86 we call distinguish 32bit/64bit follow the register name, In riscv it only has 
Xn.